--- dyfi-update.pl	2012-10-22 20:42:56.000000000 +0300
+++ new	2012-10-22 20:45:14.000000000 +0300
@@ -92,6 +92,8 @@
 use Socket;
 use strict;
 use Fcntl qw(:flock);
+use IO::Socket::SSL;
+use POSIX qw/setuid setgid/;
 
 my ($debug, $log, $pidfile, $cfgfile,
 	$update_host, $update_uri, $update_port,
@@ -113,7 +115,7 @@
 # a transparent proxy)
 $update_host = 'www.dy.fi';
 $update_uri = '/nic/update';
-$update_port = 8180;
+$update_port = 443;
 # checkip server
 $checkip_host = 'checkip.dy.fi';
 $checkip_uri = '/';
@@ -150,6 +152,10 @@
 
 ";
 
+# drop privileges
+my $username = 'dyfi';
+drop_privileges($username);
+
 #
 #	parse arguments
 #
@@ -195,6 +201,16 @@
 	exit 1;
 }
 
+# Drop privileges.
+# @param username username string
+sub drop_privileges {
+    my $username = shift;
+    
+    my ($uid, $gid) = (getpwnam($username))[2, 3];
+    setgid($gid) || crash("Can't drop privileges: $!");
+    setuid($uid) || crash("Can't drop privileges: $!");
+}
+
 #
 #	read & parse config file
 #
@@ -332,6 +348,13 @@
 		error("Cannot connect to $ht_host:$ht_port: $!");
 		return;
 	}
+
+    # turn socket to ssl socket only when credentials needed
+    if ($ht_user && $ht_passwd && !IO::Socket::SSL->start_SSL(\*SOCK)) {
+        my $err = IO::Socket::SSL::errstr();
+        error("Cannot upgrade socket to use IO::Socket::SSL: $err");
+        return;
+    }
 	
 	my($request)	= "GET $ht_uri HTTP/1.0\r\n"
 			. $basicauth
